File: python/백준/그래프/DFS/4803 트리.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

case = 1
while True:
    n, m = map(int, input().split())
    if n == 0 and m == 0:
        break

    graph = [[] for _ in range(n + 1)]
    for _ in range(m):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    visited = [False] * (n + 1)

    def check_cycle(node, parent):
        for neighbor in graph[node]:
            if neighbor == parent:
                continue
            if visited[neighbor]:
                return True          # 부모가 아닌 방문 정점이면 사이클
            visited[neighbor] = True
            if check_cycle(neighbor, node):
                return True
        return False

    tree_count = 0
    for i in range(1, n + 1):
        if not visited[i]:
            visited[i] = True
            if not check_cycle(i, 0):
                tree_count += 1

    if tree_count == 0:
        print(f"Case {case}: No trees.")
    elif tree_count == 1:
        print(f"Case {case}: There is one tree.")
    else:
        print(f"Case {case}: A forest of {tree_count} trees.")
    case += 1

File: cpp/백준/그래프/DFS/4803 트리.cpp
#include <iostream>
#include <vector>
using namespace std;

int n, m;
vector<vector<int>> graph;
vector<bool> visited;

bool check_cycle(int node, int parent) {
    for (int neighbor : graph[node]) {
        if (neighbor == parent) continue;
        if (visited[neighbor]) return true;
        visited[neighbor] = true;
        if (check_cycle(neighbor, node)) return true;
    }
    return false;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int caseNum = 1;
    while (true) {
        cin >> n >> m;
        if (n == 0 && m == 0) break;

        graph.assign(n + 1, {});
        for (int i = 0; i < m; ++i) {
            int a, b;
            cin >> a >> b;
            graph[a].push_back(b);
            graph[b].push_back(a);
        }

        visited.assign(n + 1, false);

        int tree_count = 0;
        for (int i = 1; i <= n; ++i) {
            if (!visited[i]) {
                visited[i] = true;
                if (!check_cycle(i, 0)) tree_count += 1;
            }
        }

        if (tree_count == 0) {
            cout << "Case " << caseNum << ": No trees.\n";
        } else if (tree_count == 1) {
            cout << "Case " << caseNum << ": There is one tree.\n";
        } else {
            cout << "Case " << caseNum << ": A forest of " << tree_count << " trees.\n";
        }
        caseNum += 1;
    }
    return 0;
}
